libcalico-go/lib/metricsserver/metrics_server.go

```diff
@@ -35,7 +38,101 @@ func ServePrometheusMetricsForever(host string, port int) {
 		}).Info("Starting prometheus metrics endpoint")
 		err := http.ListenAndServe(addr, mux)
 		logrus.WithError(err).Error(
-			"Prometheus metrics endpoint failed, trying to restart it...")
+			"Prometheus http metrics endpoint failed, trying to restart it...")
 		time.Sleep(1 * time.Second)
 	}
 }
+
+// ServePrometheusMetricsHTTPS starts a secure Prometheus metrics server with dynamic TLS certificate reloading.
+func ServePrometheusMetricsHTTPS(host string, port int, certFile, keyFile, minTLSVersion, clientAuthType, caFile string) {
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.Handler())
+	addr := fmt.Sprintf("[%v]:%v", host, port)
+
+	// Initial TLS config loading to catch errors early.
+	tlsConfig, err := loadTLSConfig(certFile, keyFile, minTLSVersion, clientAuthType, caFile)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to load initial TLS configuration")
+	}
+
+	// Enable dynamic certificate reloading.
+	tlsConfig.GetConfigForClient = func(*tls.ClientHelloInfo) (*tls.Config, error) {
+		return loadTLSConfig(certFile, keyFile, minTLSVersion, clientAuthType, caFile)
+	}
+
+	server := &http.Server{
+		Addr:      addr,
+		TLSConfig: tlsConfig,
+		Handler:   mux,
+	}
+
+	// Restart server on failure.
+	for {
+		logrus.WithFields(logrus.Fields{
+			"host": host,
+			"port": port,
+		}).Info("Starting Prometheus metrics endpoint with TLS")
+
+		err = server.ListenAndServeTLS("", "")
+		if err != nil {
+			logrus.WithError(err).Error("Prometheus https metrics endpoint failed, restarting...")
+			time.Sleep(1 * time.Second)
+		}
+	}
+}
+
+// loadTLSConfig dynamically loads the TLS certificates and keys.
+func loadTLSConfig(certFile, keyFile, tlsMinVersion, clientAuthType, caFile string) (*tls.Config, error) {
```

Comment:
It would be nice if we could leverage the `github.com/projectcalico/calico/crypto/pkg/tls` package for this (either importing that to generate the base config, or moving this logic into that package?)